core/serialization/deserializer/montage-reviver.js

```diff
@@ -469,14 +465,20 @@ var MontageReviver = exports.MontageReviver = Montage.specialize(/** @lends Mont
                 )) {
                 // We have a circular reference. If we wanted to forbid circular
                 // references this is where we would throw an error.
-                return Promise.resolve(this._deserializerConstructor.moduleContexts.get(location)._objects.root);
+                return this._deserializerConstructor.moduleContexts.get(location)._objects.root;
             }
 
             if (isObjectDescriptor && !Promise.is(module) && !module.montageObject) {
                 module = context._require.async(locationDesc.moduleId);
             }
 
             if (Promise.is(module)) {
+                if (this._sync) {
+                    throw new Error(
+                        "Tried to revive montage object with label " + label +
+                        " synchronously but the module was not loaded: " + JSON.stringify(value)
+                    );
+                }
                 return module.then(function (exports) {
```

Comment:
How can this still be async if the reviver is sync?

Reply:
I don't understand the question. The intention here is that if we are sync but the module we are loading is async, we need to throw an error.

Reply:
I don't understand the question. The intention here is that if we are sync but the module we are loading is async, we need to throw an error.

Reply:
Doesn't context._require.async(locationDesc.moduleId); returns a promise anyway? Shouldn't require() be used if sync instead?

Reply:
Yes. Looks like the condition on 471 means the module will be loaded async if it is an .mjson but has not been compiled into a `montageObject` yet. If we are sync and we fall into that condition, it means our caller did not take care of loading all of our dependencies beforehand, and we should throw.

Reply:
@marchant Ok with this?

Reply:
yep, merged it!